### PR TITLE
Remove flattening observation in samplers

### PR DESCRIPTION
**Pull request by ahtsan**
Flattening of observations should be decoupled to the consumer, e.g. algorithms. The assumption that the observations should be flattened is not valid in some use cases, e.g. when observations are images.

### CODE REVIEW

**Comment by ryanjulian** [comment]
Don't you need to add the flattening back into process_samples? Did you check that the on-policy algos didn't break?

**Comment by ahtsan** [comment]
> Don't you need to add the flattening back into process_samples? Did you check that the on-policy algos didn't break?

All tests passed except `test_ray_batch_sampler`, since it requires one-hot representation for `Discrete` observation space. Do we need to add flattening back?

**Comment by ryanjulian** [comment]
hmm is that a bug in RaySampler? What does `Discrete.sample()` return according to its API?

**Comment by ryanjulian** [comment]
Please train at least one on-policy algorithm on one environment fully to make sure this is actually working. Make sure the environment has multi-dimensional observation space (otherwis `[obs] == flatten_n([obs])` and we're not testing anything :)

**Comment by ahtsan** [comment]
> hmm is that a bug in RaySampler? What does `Discrete.sample()` return according to its API?

`env.sample()` returns a scalar value between 0 and n for a `env=Discrete(n)`. In garage we were calling `Discrete.flatten()` to make it one-hot.

**Comment by ahtsan** [comment]
> Please train at least one on-policy algorithm on one environment fully to make sure this is actually working. Make sure the environment has multi-dimensional observation space (otherwis `[obs] == flatten_n([obs])` and we're not testing anything :)

Do you mean like `Box(m, n)`?

**Comment by ahtsan** [comment]
Categorical MLP Policy 3 trials on

Assault-ramDeterminstic
Assault-ramDeterministic-v4_benchmark_mean

CartPole
CartPole-v1_benchmark_mean

LunarLander
LunarLander-v2_benchmark_mean

**Comment by ahtsan** [comment]
@ryanjulian ready for review - basically I move the flattening to `batch_poplot` and flatten the obs and actions according to a boolean flag.

File: src/garage/tf/misc/tensor_utils.py [hunk 1]
```diff
@@ -48,32 +53,50 @@ def get_target_ops(variables, target_variables, tau=None):
...
     return tf.reshape(t, [-1] + list(t.shape[2:]), name=name)
 
 
 def flatten_batch_dict(d, name=None):
+    """Flatten batch dict."""
     with tf.name_scope(name, 'flatten_batch_dict', [d]):
         return {k: flatten_batch(v) for k, v in d.items()}
 
 
+def flatten_input_along_dim(t, dim, name='flatten_input_full'):
```
Comment (ryanjulian) [inline src/garage/tf/misc/tensor_utils.py]: how is this different from flatten_batch?

please include a full docstring

File: tests/integration_tests/test_flattening.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from garage.tf.envs import TfEnv
+from garage.tf.experiment import LocalTFRunner
+from garage.tf.policies import CategoricalMLPPolicy
+from tests.fixtures import snapshot_config, TfGraphTestCase
+from tests.fixtures.envs.wrappers import ReshapeObservation
+
+
+class TestResume(TfGraphTestCase):
+
+    def test_ppo_pendulum_flatten_input(self):
```
Comment (ryanjulian) [inline tests/integration_tests/test_flattening.py]: maybe just put this along with all the other PPO tests?

File: src/garage/tf/algos/npo.py [hunk 3]
```diff
@@ -214,17 +224,20 @@ def _build_inputs(self):
...
+                                            shape=[None] * 2 + list(shape),
+                                            name='policy_old_%s' % k)
                 for k, shape in policy_dist.dist_info_specs
             }
             policy_old_dist_info_vars_list = [
                 policy_old_dist_info_vars[k]
                 for k in policy_dist.dist_info_keys
             ]
 
+            if self.flatten_input:
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: can you write out some comments here describing the shapes of the inputs and their flattened representation? 

how does this case contrast with the one immediately below it, which flattens the observation? why is the flattening done twice?

**Comment by ryanjulian** [comment]
It will probably be simpler to merge this change if you move the flattening to process_samples, rather than trying to do it in TF.

NPO probably needs a bit of a refactor with how it deals with flattening and valid time steps. This change just makes it more confusing.

If you do it in process samples, you can just move the function call over from the sampler and not worry about new functions, tests, documentation, etc.

File: src/garage/tf/algos/npo.py (hunk 3 above)
Comment (ryanjulian) [reply src/garage/tf/algos/npo.py]: why is this step necessary if you already flattened the input in `process_samples`?

Comment (ahtsan) [reply src/garage/tf/algos/npo.py]: In `process_samples` the samples (training data / real values) are flattened, while the flattening here is for the placeholders.

File: tests/integration_tests/test_flattening.py (hunk 2 above)
Comment (ahtsan) [reply tests/integration_tests/test_flattening.py]: sure

File: src/garage/tf/algos/npo.py (hunk 3 above)
Comment (ryanjulian) [reply src/garage/tf/algos/npo.py]: ah. i think it would be more clear if you did this at placeholder creation-time.

File: src/garage/tf/misc/tensor_utils.py (hunk 1 above)
Comment (ahtsan) [reply src/garage/tf/misc/tensor_utils.py]: Will do.
In short, `flatten_batch` flattens the batch dimension while keeping the sample dimension, and this function is able to flatten the observation dimension, when the given `dim` is 2 (since we have [batch_dim, time_step, observation_dim] for samples in garage).

File: src/garage/tf/algos/npo.py (hunk 3 above)
Comment (ryanjulian) [reply src/garage/tf/algos/npo.py]: ah -- this would be much more clear if it were placed near placeholder creation and generously commented.

is there a way we can make the placeholder this dimension in the first place, rather than post-processing it?

Comment (ahtsan) [reply src/garage/tf/algos/npo.py]: I thought it would be more straight forward if we put the boolean flag in `akro.space.to_tf_placeholder()`?

Comment (ryanjulian) [reply src/garage/tf/algos/npo.py]: yes -- but i'm trying to avoid bumping the akro version number. let's see if this feature is needed in more than one algorithm first.

remember the goal here is to get models in ASAP. NPO is gonna get rewritten to use tf.distributions soon, and that might let us get rid of (at least one kind of) some of the flattening.

Comment (ahtsan) [reply src/garage/tf/algos/npo.py]: How about we do minimal changes - pass the boolean flag to algo and do the checking in sampler by accessing `self.algo.is_flatten_input`, and flatten if the flag is on? We can move the checking back to algo when we rewrite later. Agree that the goal is to get models in ASAP.

Comment (ryanjulian) [reply src/garage/tf/algos/npo.py]: that change may be minimal but it introduces a bunch of technical debt (and now you need to test a new case in the sampler).

again, why can't the algorithm just have a flag which lets it decide whether to flatten the observations?

today:
1. sampler produces flattened observations
1.  algo unconditionally flattens observations in `process_samples`
1. `optimize_policy` expects flattened observations (e.g. placeholders are flat)

tomorrow:
1. samples produces unflattened observations
1.  algo conditionally flattens observations in `process_samples`
1. `optimize_policy` checks whether to expect flattened or unflattened observations and actions.

it seems you are stuck on step (3) of "tomorrow" complexity-wise. rather than adding a whole new function to flatten your placeholder, why not infer its shape directly from the space? you know the [batch, time] dimensions already anyway.

you can file an issue in akro to add a flattening option in the tensor creation API.

```python
if self.flatten_inputs:
    obs_var = tf.compat.v1.placeholder(
        name='obs',
        dtype=tf.float32,
        shape=[None, None, self.observation_space.flat_dim])
else:
    obs_var = self.observation_space.to_tf_placeholder(name='obs', batch_dims=2)
```

Comment (ahtsan) [reply src/garage/tf/algos/npo.py]: It's a good idea, and we can avoid the new function `flatten_input_along_dim`. Will do.

**Comment by ahtsan** [comment]
@ryanjulian Please see the new changes.

File: src/garage/tf/algos/npo.py [hunk 4]
```diff
@@ -98,6 +99,7 @@ def __init__(self,
                  use_neg_logli_entropy=False,
                  stop_entropy_gradient=False,
                  entropy_method='no_entropy',
+                 flatten_input=True,
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: you need to update the docstring

**Comment by ryanjulian** [comment]
you might need to remove your changes to tensor_utils to get past codecov.

File: src/garage/tf/algos/batch_polopt.py [hunk 5]
```diff
@@ -138,6 +139,18 @@ def process_samples(self, itr, paths):
 
         max_path_length = self.max_path_length
 
+        if self.flatten_input:
+            paths = list(
```
Comment (krzentner) [inline src/garage/tf/algos/batch_polopt.py]: This would definitely be clearer as a list comprehension.

**Comment by ryanjulian** [comment]
See the "Note": https://www.tensorflow.org/versions/r1.14/api_docs/python/tf/keras/layers/Dense

Seems like the flattening might not even be necessary in the future.

File: src/garage/tf/algos/batch_polopt.py (hunk 5 above)
Comment (ahtsan) [reply src/garage/tf/algos/batch_polopt.py]: @krzentner fixed
